File: backend/src/common/iraq-geo.ts
import { City, Governorate } from '@prisma/client';

/**
 * Which cities belong to which governorate. Used to validate that an address's
 * city is consistent with its governorate, and to drive the dependent city
 * dropdown on the frontend (mirrored in frontend/src/lib/iraqGeo.ts).
 */
export const GOVERNORATE_CITIES: Record<Governorate, City[]> = {
  BAGHDAD: [City.BAGHDAD, City.ABU_GHRAIB, City.MAHMUDIYAH, City.TARMIYAH, City.SADR_CITY],
  BASRA: [City.BASRA, City.ZUBAIR, City.ABU_AL_KHASIB, City.QURNA, City.FAW],
  NINEVEH: [City.MOSUL, City.TAL_AFAR, City.SINJAR, City.HAMDANIYA, City.SHEIKHAN],
  ERBIL: [City.ERBIL, City.SHAQLAWA, City.KOYA, City.SORAN, City.CHOMAN],
  SULAYMANIYAH: [City.SULAYMANIYAH, City.RANIA, City.CHAMCHAMAL, City.KALAR, City.DUKAN],
  DUHOK: [City.DUHOK, City.ZAKHO, City.AMEDI, City.SEMEL, City.AKRE],
  KIRKUK: [City.KIRKUK, City.HAWIJA, City.DAQUQ, City.DIBIS],
  NAJAF: [City.NAJAF, City.KUFA, City.MISHKHAB, City.MANATHERA],
  KARBALA: [City.KARBALA, City.AIN_AL_TAMR, City.HINDIYA],
  BABYLON: [City.HILLAH, City.MUSAYYIB, City.MAHAWIL, City.HASHIMIYA],
  WASIT: [City.KUT, City.SUWAIRA, City.NUMANIYA, City.AZIZIYA, City.BADRA],
  MAYSAN: [City.AMARAH, City.MAJAR_AL_KABIR, City.ALI_AL_GHARBI, City.QALAT_SALIH],
  DHI_QAR: [City.NASIRIYAH, City.SHATRA, City.RIFAI, City.SUQ_AL_SHUYUKH, City.CHIBAYISH],
  MUTHANNA: [City.SAMAWAH, City.RUMAITHA, City.KHIDHIR, City.SALMAN],
  QADISIYYAH: [City.DIWANIYAH, City.AFAK, City.SHAMIYA, City.HAMZA],
  DIYALA: [City.BAQUBAH, City.MUQDADIYAH, City.KHALIS, City.KHANAQIN, City.BALADRUZ],
  ANBAR: [City.RAMADI, City.FALLUJAH, City.HIT, City.HADITHA, City.QAIM, City.RUTBA],
  SALADIN: [City.TIKRIT, City.SAMARRA, City.BALAD, City.DUJAIL, City.SHIRQAT, City.BAIJI],
  HALABJA: [City.HALABJA, City.KHURMAL, City.SAYID_SADIQ, City.BIYARA],
};

/** True when `city` is a valid city of `governorate`. */
export const cityInGovernorate = (governorate: Governorate, city: City): boolean =>
  GOVERNORATE_CITIES[governorate]?.includes(city) ?? false;

File: mobile/lib/features/addresses/domain/iraq_geo.dart
/// Iraq governorate → city mapping, mirrored from the backend
/// (`backend/src/common/iraq-geo.ts`). Drives the dependent city dropdown on the
/// address form and keeps the client from submitting an invalid pair (the
/// backend re-validates regardless). Values are the Prisma enum names sent to
/// the API verbatim; [humanizeGeo] turns them into display labels.
class IraqGeo {
  IraqGeo._();

  static const Map<String, List<String>> governorateCities = {
    'BAGHDAD': ['BAGHDAD', 'ABU_GHRAIB', 'MAHMUDIYAH', 'TARMIYAH', 'SADR_CITY'],
    'BASRA': ['BASRA', 'ZUBAIR', 'ABU_AL_KHASIB', 'QURNA', 'FAW'],
    'NINEVEH': ['MOSUL', 'TAL_AFAR', 'SINJAR', 'HAMDANIYA', 'SHEIKHAN'],
    'ERBIL': ['ERBIL', 'SHAQLAWA', 'KOYA', 'SORAN', 'CHOMAN'],
    'SULAYMANIYAH': ['SULAYMANIYAH', 'RANIA', 'CHAMCHAMAL', 'KALAR', 'DUKAN'],
    'DUHOK': ['DUHOK', 'ZAKHO', 'AMEDI', 'SEMEL', 'AKRE'],
    'KIRKUK': ['KIRKUK', 'HAWIJA', 'DAQUQ', 'DIBIS'],
    'NAJAF': ['NAJAF', 'KUFA', 'MISHKHAB', 'MANATHERA'],
    'KARBALA': ['KARBALA', 'AIN_AL_TAMR', 'HINDIYA'],
    'BABYLON': ['HILLAH', 'MUSAYYIB', 'MAHAWIL', 'HASHIMIYA'],
    'WASIT': ['KUT', 'SUWAIRA', 'NUMANIYA', 'AZIZIYA', 'BADRA'],
    'MAYSAN': ['AMARAH', 'MAJAR_AL_KABIR', 'ALI_AL_GHARBI', 'QALAT_SALIH'],
    'DHI_QAR': ['NASIRIYAH', 'SHATRA', 'RIFAI', 'SUQ_AL_SHUYUKH', 'CHIBAYISH'],
    'MUTHANNA': ['SAMAWAH', 'RUMAITHA', 'KHIDHIR', 'SALMAN'],
    'QADISIYYAH': ['DIWANIYAH', 'AFAK', 'SHAMIYA', 'HAMZA'],
    'DIYALA': ['BAQUBAH', 'MUQDADIYAH', 'KHALIS', 'KHANAQIN', 'BALADRUZ'],
    'ANBAR': ['RAMADI', 'FALLUJAH', 'HIT', 'HADITHA', 'QAIM', 'RUTBA'],
    'SALADIN': ['TIKRIT', 'SAMARRA', 'BALAD', 'DUJAIL', 'SHIRQAT', 'BAIJI'],
    'HALABJA': ['HALABJA', 'KHURMAL', 'SAYID_SADIQ', 'BIYARA'],
  };

  static List<String> get governorates => governorateCities.keys.toList();

  static List<String> citiesOf(String? governorate) =>
      governorateCities[governorate] ?? const [];

  static bool isValidPair(String? governorate, String? city) =>
      citiesOf(governorate).contains(city);
}

/// Turns an enum name (`ABU_AL_KHASIB`) into a readable label (`Abu Al Khasib`).
/// Place names read the same across the app's languages, so a single
/// transliterated label is used for all locales.
String humanizeGeo(String value) => value
    .split('_')
    .map((w) => w.isEmpty ? w : '${w[0]}${w.substring(1).toLowerCase()}')
    .join(' ');
